ddl/ddl_api.go

```diff
@@ -1424,23 +1425,34 @@ func buildTableInfo(
 			if err != nil {
 				return nil, err
 			}
-			if !config.GetGlobalConfig().AlterPrimaryKey {
-				singleIntPK := isSingleIntPK(constr, lastCol)
-				clusteredIdx := ctx.GetSessionVars().EnableClusteredIndex
-				if singleIntPK || clusteredIdx {
-					// Primary key cannot be invisible.
-					if constr.Option != nil && constr.Option.Visibility == ast.IndexVisibilityInvisible {
-						return nil, ErrPKIndexCantBeInvisible
-					}
-				}
-				if singleIntPK {
+			pkTp, alterPKConf := model.PrimaryKeyTypeDefault, config.GetGlobalConfig().AlterPrimaryKey
+			if constr.Option != nil {
+				pkTp = constr.Option.PrimaryKeyTp
+			}
+			switch pkTp {
+			case model.PrimaryKeyTypeNonClustered:
+				break
+			case model.PrimaryKeyTypeClustered:
+				if isSingleIntPK(constr, lastCol) {
 					tbInfo.PKIsHandle = true
-					// Avoid creating index for PK handle column.
-					continue
-				}
-				if clusteredIdx {
+				} else {
 					tbInfo.IsCommonHandle = true
 				}
+			case model.PrimaryKeyTypeDefault:
+				if isSingleIntPK(constr, lastCol) {
+					tbInfo.PKIsHandle = !alterPKConf
```

Comment:
Add `&& ctx.GetSessionVars().EnableClusteredIndex` either, to make it consistent with common handle?

Reply:
A lot of tests will be broken. Let's do this in another PR when it is decided.